google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ConnectionOptions.java

```diff
@@ -537,22 +555,22 @@ private ConnectionOptions(Builder builder) {
     this.credentialsUrl =
         builder.credentialsUrl != null ? builder.credentialsUrl : parseCredentials(builder.uri);
     this.encodedCredentials = parseEncodedCredentials(builder.uri);
-    // Check that not both a credentials location and encoded credentials have been specified in the
-    // connection URI.
-    Preconditions.checkArgument(
-        this.credentialsUrl == null || this.encodedCredentials == null,
-        "Cannot specify both a credentials URL and encoded credentials. Only set one of the properties.");
-
+    this.credentialsProvider = parseCredentialsProvider(builder.uri);
     this.oauthToken =
         builder.oauthToken != null ? builder.oauthToken : parseOAuthToken(builder.uri);
-    this.fixedCredentials = builder.credentials;
-    // Check that not both credentials and an OAuth token have been specified.
+    // Check that at most one of credentials location, encoded credentials, credentials provider and
+    // OUAuth token has been specified in the connection URI.
     Preconditions.checkArgument(
-        (builder.credentials == null
-                && this.credentialsUrl == null
-                && this.encodedCredentials == null)
-            || this.oauthToken == null,
-        "Cannot specify both credentials and an OAuth token.");
+        Stream.of(
```

Comment:
Why did we remove `builder.credentials` (or `this.fixedCredentials`) from the stream here, and move the fixedCredentials initializations to after this nonNull Preconditions check?

Reply:
`builder.credentials` is not part of the public API and is not something users can set, so we don't need to check it here. (see https://github.com/googleapis/java-spanner/blob/2360e7e1bfd758fd773d52b43f31c20b41d77c97/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ConnectionOptions.java#L495)

Reply:
Thanks for clarifying!